subprojects/base-services/src/main/java/org/gradle/internal/classloader/TransformReplacer.java

```diff
@@ -185,26 +192,40 @@ public synchronized byte[] loadTransformedClass(String className) throws IOExcep
 
         @Override
         public synchronized void close() {
+            // Not calling getJarFileLocked intentionally, to avoid opening the JAR if it isn't opened yet.
             IoActions.closeQuietly(jarFile);
         }
 
-        @SuppressWarnings("Since15")
         private JarFile getJarFileLocked() throws IOException {
             ensureOpened();
             if (jarFile == null) {
-                try {
-                    // Set up the MR-JAR properly when running on Java 9+.
-                    jarFile = new JarFile(jarFilePath, true, JarFile.OPEN_READ, JarFile.runtimeVersion());
-                } catch (NoSuchMethodError e) {
-                    // Running on Java 8, fall back to the old ways.
-                    jarFile = new JarFile(jarFilePath);
+                jarFile = JarCompat.open(jarFilePath);
+                if (jarFile.isMultiRelease() && !isTransformed(jarFile.getJarFile())) {
+                    throw new GradleException(String.format(
+                        "Cannot load multi-release JAR '%s' because it cannot be fully instrumented for Java %d by this version of Gradle. Please use supported Java version (<=%d).",
+                        originalJarFilePath.getAbsolutePath(),
+                        jarFile.javaRuntimeVersionUsed(),
+                        AsmConstants.MAX_SUPPORTED_JAVA_VERSION));
                 }
             }
-            return jarFile;
+            return jarFile.getJarFile();
         }
 
         private String classNameToPath(String className) {
             return className + ".class";
         }
     }
+
+    private static boolean isTransformed(JarFile jarFile) throws IOException {
+        JarEntry entry = jarFile.getJarEntry(MARKER_RESOURCE_NAME);
+        if (entry != null) {
+            InputStream in = jarFile.getInputStream(entry);
+            try {
+                return "true".equals(StreamByteBuffer.of(in).readAsString("UTF-8"));
```

Comment:
💡 Since this marker is only ever read by Gradle, why not make it a single byte?

Reply:
Thanks! In fact, we can go deeper and make it 0 bytes for the common case, and only use a non-empty file for the unsupported versions.

Reply:
Even better! 😃